File: src/test/java/com/letgo/book/unit/application/changeTitle/ChangeTitleCommandHandlerTest.java
package com.letgo.book.unit.application.changeTitle;

import com.letgo.book.application.changeTitle.ChangeTitleCommand;
import com.letgo.book.application.changeTitle.ChangeTitleCommandHandler;
import com.letgo.book.domain.*;
import com.letgo.book.unit.application.BookTestCase;
import com.letgo.book.unit.domain.BookTitleMother;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

final public class ChangeTitleCommandHandlerTest extends BookTestCase {
    private final ChangeTitleCommandHandler handler = new ChangeTitleCommandHandler(
            repository,
            new BookFinder(repository),
            publisher
    );

    @Test
    public void itShouldChangeAnExistingBookTitle() {
        Book currentBook = anExistingBook();
        BookId id = currentBook.id();

        BookTitle newTitle = BookTitleMother.random();

        expectDomainEventsToBePublished(new BookTitleChanged(
                id.value(),
                currentBook.title().value(),
                newTitle.value(),
                newTitle.createdAt()
        ));

        handler.handle(new ChangeTitleCommand(id.value(), newTitle.value(), newTitle.createdAt().toString()));

        Book book = repository.find(id).orElseThrow();
        assertEquals(id, book.id());
        assertEquals(newTitle, book.title());

        eventsShouldBePublished();
    }

    @Test
    public void itShouldBeIdempotent() {
        Book currentBook = anExistingBook();
        BookId id = currentBook.id();
        BookTitle title = currentBook.title();

        expectDomainEventsToBePublished();

        handler.handle(new ChangeTitleCommand(id.value(), title.value(), title.createdAt().toString()));

        assertEquals(currentBook, repository.find(id).orElseThrow());
        eventsShouldNotBePublished();
    }
}

File: src/test/java/com/letgo/book/unit/application/changeTitle/ChangeTitleCommandHandlerTest.kt
package com.letgo.book.unit.application.changeTitle

import com.letgo.book.application.changeTitle.ChangeTitleCommand
import com.letgo.book.application.changeTitle.ChangeTitleCommandHandler
import com.letgo.book.domain.BookFinder
import com.letgo.book.domain.BookStatus
import com.letgo.book.domain.BookTitleChanged
import com.letgo.book.unit.application.BookTestCase
import com.letgo.book.unit.domain.ABookTitle
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

private class ChangeTitleCommandHandlerTest : BookTestCase() {
    private val handler = ChangeTitleCommandHandler(repository, BookFinder(repository), publisher)

    @Test
    fun `It should change an existing book title`() {
        val currentBook = anExistingBook()
        val currentBookId = currentBook.id()
        val newTitle = ABookTitle.random()
        val updatedBook = currentBook.copy(
            title = newTitle,
            status = BookStatus.Edited,
        )

        expectDomainEventsToBePublished(
            BookTitleChanged(
                aggregateId = currentBookId.value(),
                oldTitle = currentBook.title().value(),
                newTitle = newTitle.value(),
                occurredOn = newTitle.createdAt(),
            )
        )

        handler.handle(
            ChangeTitleCommand(
                id = currentBookId.value(),
                newTitle = newTitle.value(),
                occurredOn = newTitle.createdAt().toString(),
            )
        )

        assertEquals(updatedBook, repository.find(currentBookId))

        eventsShouldBePublished()
    }

    @Test
    fun `It should not apply earlier changes`() {
        val currentBook = anExistingBook()
        val currentBookId = currentBook.id()
        val currentBookTitle = currentBook.title()
        val earlierTitle = ABookTitle.with(
            title = currentBook.title().value() + "olakease",
            createdAt = currentBookTitle.createdAt().minusHours(1),
        )

        expectDomainEventsToBePublished()

        handler.handle(
            ChangeTitleCommand(
                id = currentBookId.value(),
                newTitle = earlierTitle.value(),
                occurredOn = earlierTitle.createdAt().toString(),
            )
        )

        assertEquals(currentBook, repository.find(currentBookId))

        eventsShouldNotBePublished()
    }

    @Test
    fun `It should be idempotent`() {
        val currentBook = anExistingBook()
        val currentBookId = currentBook.id()
        val currentTitle = currentBook.title()

        expectDomainEventsToBePublished()

        handler.handle(
            ChangeTitleCommand(
                id = currentBookId.value(),
                newTitle = currentTitle.value(),
                occurredOn = currentTitle.createdAt().toString(),
            )
        )

        assertEquals(currentBook, repository.find(currentBookId))

        eventsShouldNotBePublished()
    }
}
